### PR TITLE
RDMA/hfi1: Fix use-after-free bug for mm struct

### PR DESCRIPTION
**Pull request by drmiller-cni**
Under certain conditions, such as MPI_Abort, the hfi1 cleanup
code may represent the last reference held on the task mm.
hfi1_mmu_rb_unregister() then drops the last reference and the mm is
freed before the final use in hfi1_release_user_pages().  A new task
may allocate the mm structure while it is still being used, resulting in
problems. One manifestation is corruption of the mmap_sem counter leading
to a hang in down_write().  Another is corruption of an mm struct that
is in use by another task.

One frequently-seen hang shows stack traces like:

schedule+0x29/0x70
rwsem_down_write_failed+0x215/0x3c0
call_rwsem_down_write_failed+0x17/0x30
down_write+0x2d/0x3d
hfi1_release_user_pages+0x4b/0x90 [hfi1]
cacheless_tid_rb_remove+0x10e/0x260 [hfi1]
tid_rb_remove+0xe/0x10 [hfi1]
hfi1_mmu_rb_unregister+0xed/0x150 [hfi1]
hfi1_user_exp_rcv_free+0x39/0xf0 [hfi1]
hfi1_file_close+0x8b/0x1c0 [hfi1]
__fput+0xec/0x230
____fput+0xe/0x10
task_work_run+0xbb/0xe0
do_exit+0x2d4/0xa30
do_group_exit+0x3f/0xa0
get_signal_to_deliver+0x1ce/0x5e0
do_signal+0x57/0x6f0
do_notify_resume+0x72/0xc0
retint_signal+0x48/0x8c

INTERNAL: STL-63023: Fix mm use after free

Fixes: 3d2a9d642512 ("IB/hfi1: Ensure correct mm is used at all times")
Cc: <stable@vger.kernel.org>
Signed-off-by: Douglas Miller <doug.miller@cornelisnetworks.com>

### CODE REVIEW

**Comment by drmiller-cni** [comment]
Test results here: http://aw-fi-jenkins-01.cornelisnetworks.com/view/github%20kmod%20testing/job/STL1_WFR-special_testing_RHEL7.8/10/.

Attaching transcript of build here: 
[opa-hfi1-build-1.txt](https://github.com/cornelisnetworks/opa-hfi1/files/8451389/opa-hfi1-build-1.txt)

File: drivers/infiniband/hw/hfi1/mmu_rb.c [hunk 1]
```diff
@@ -157,25 +157,23 @@ void hfi1_mmu_rb_unregister(struct mmu_rb_handler *handler)
...
+	 * so we keep one mmgrab here and do the mmdrop later.
+	 * Except for the !NO_NOTIFIER_REG_GRAB case, we already
+	 * have an extra mmgrab at this point.
+	 */
 	if (handler->registered) {
 #ifdef NO_MMU_NOTIFIER_MM
-		mmu_notifier_unregister(&handler->mn, handler->mm);
-#else
-		mmu_notifier_unregister(&handler->mn, handler->mn.mm);
+#ifndef NO_NOTIFIER_REG_GRAB
```
Comment (ddalessa) [inline drivers/infiniband/hw/hfi1/mmu_rb.c]: I think this basically needs to be the same as the drop logic. We always want to do a grab/ungrab. That's the fix right? So the only difference between distros is if the mm is in the notifier or directly in our handler.

File: drivers/infiniband/hw/hfi1/mmu_rb.c [hunk 2]
```diff
@@ -157,25 +157,23 @@ void hfi1_mmu_rb_unregister(struct mmu_rb_handler *handler)
...
-		mmdrop(handler->mm);
+		mmu_notifier_unregister(&handler->mn, handler->mm);
 #else
-		mmdrop(handler->mn.mm);
+#ifndef NO_NOTIFIER_REG_GRAB
+		mmgrab(handler->mn.mm);
 #endif
-#endif
-	} else {
-		/* The grab was unconditional for the GPU case */
```
Comment (ddalessa) [inline drivers/infiniband/hw/hfi1/mmu_rb.c]: Not sure about the GPU angle here.

File: drivers/infiniband/hw/hfi1/mmu_rb.c (hunk 1 above)
Comment (drmiller-cni) [reply drivers/infiniband/hw/hfi1/mmu_rb.c]: We don't NEED to do the mmgrab unless the only reference we have was by mmu_notifier_register. So, we could unconditionally do an mmgrab before any of the unregister stuff, and leave the original conditional mmdrop. The resulting code would be a little odd/sub-optimal in that it does mmgrab, mmdrop, mmdrop. But, might be clearer as far as diffs are concerned. Would you rather do it that way then?

Comment (ddalessa) [reply drivers/infiniband/hw/hfi1/mmu_rb.c]: So NO_NOTIFER_REG_GRAB means we did an explicit mmgrab in our code. So if that's NOT on then we have to do the grab here. Ok makes sense.

**Review by ddalessa — APPROVED** [review]

**Comment by drmiller-cni** [comment]
Second version of patch build transcript is attached.
[opa-hfi1-build-2.txt](https://github.com/cornelisnetworks/opa-hfi1/files/8457070/opa-hfi1-build-2.txt)

Test run results are at http://aw-fi-jenkins-01.cornelisnetworks.com/view/github%20kmod%20testing/job/STL1_WFR-special_testing_RHEL7.8/12/. The overall result is fail but the test results all say PASSED. It appears that the BMC console connection was lost on the first node, which may be the reason for the FAILED status.

**Pull request merged**